src/cis4.rs

```diff
@@ -0,0 +1,259 @@
+//! This module contains types and functions for interacting with smart
+//! contracts following the [CIS-4](https://proposals.concordium.software/CIS/cis-4.html) specification.
+//!
+//! The type [`Cis4Contract`](crate::cis4::Cis4Contract) acts as a wrapper
+//! around the [Client](crate::v2::Client) and a contract address providing
+//! functions for querying and making transactions to smart contract.
+
+use crate::{
+    contract_client::*,
+    types::{transactions, RejectReason},
+    v2::IntoBlockIdentifier,
+};
+pub use concordium_base::{cis2_types::MetadataUrl, cis4_types::*};
+use concordium_base::{
+    constants::MAX_PARAMETER_LEN,
+    contracts_common::{self, AccountAddress},
+    hashes::TransactionHash,
+    smart_contracts::{ExceedsParameterSize, OwnedParameter},
+    web3id::{CredentialHolderId, Web3IdSigner, REVOKE_DOMAIN_STRING},
+};
+
+#[derive(thiserror::Error, Debug)]
+/// An error that can occur when executing CIS4 queries.
+pub enum Cis4QueryError {
+    /// The smart contract receive name is invalid.
+    #[error("Invalid receive name: {0}")]
+    InvalidReceiveName(#[from] contracts_common::NewReceiveNameError),
+
+    /// A general RPC error occured.
+    #[error("RPC error: {0}")]
+    RPCError(#[from] super::v2::QueryError),
+
+    /// The data returned from q query could not be parsed.
+    #[error("Failed parsing the response.")]
+    ResponseParseError(#[from] contracts_common::ParseError),
+
+    /// The node rejected the invocation.
+    #[error("Rejected by the node: {0:?}.")]
+    NodeRejected(crate::types::RejectReason),
+}
+
+impl From<RejectReason> for Cis4QueryError {
+    fn from(value: RejectReason) -> Self { Self::NodeRejected(value) }
+}
+
+#[derive(thiserror::Error, Debug)]
+/// An error that can occur when sending CIS4 update transactions.
+pub enum Cis4TransactionError {
+    /// The smart contract receive name is invalid.
+    #[error("Invalid receive name: {0}")]
+    InvalidReceiveName(#[from] contracts_common::NewReceiveNameError),
+
+    /// The parameter is too large.
+    #[error("Parameter is too large: {0}")]
+    InvalidParams(#[from] ExceedsParameterSize),
+
+    /// A general RPC error occured.
+    #[error("RPC error: {0}")]
+    RPCError(#[from] super::v2::RPCError),
+
+    /// The node rejected the invocation.
+    #[error("Rejected by the node: {0:?}.")]
+    NodeRejected(crate::types::RejectReason),
+}
+
+/// Transaction metadata for CIS-4 update transactions.
+pub type Cis4TransactionMetadata = ContractTransactionMetadata;
+
+#[doc(hidden)]
+#[derive(Debug, Clone, Copy)]
+pub enum Cis4Type {}
```

Comment:
This should probably be visible in the documentation and explained.

Reply:
added